### DISCUSSION TITLE
Offline feature

### DISCUSSION THREAD

**Discussion by nastiliano** [sqlitecloud/discussions#22] [General]
Hello everybody,

I think this project is very interesting.

I felt in love just when I saw the "Offline-First" paradigm for SQLite about 6 months ago and I've been waiting for long to try this capability, but months passing and this feature is not announced.

First I waited to try the platform and then I saw that this functionality was not enabled... The platform seems spectacular but without this feature I can't still going straight in my project, in fact I can't see difference about a MariaDB database and a SQLite Cloud database (for now).

Last information that I have is this feature will be avaliable at the end of this month (in other words, now),

Somebody have information about that?

Thank you very much.

Best regards.

**Comment by btholt** [comment]
@nastiliano Still in progress! We're looking at a more August/September timeframe for people to be able to try it out.

**Reply by nastiliano** [reply]
Thanks for your reply.

But now, the million dollar question:

Can I program my app with Python SQLite Cloud SDK as usually and you assure me that in august/september we'll be able to enjoy this feature "automatically"?

Thank you very much.
Best Regards

**Reply by btholt** [reply]
We're striving to make this feature as easy and transparent as possible. Dev experience (and your customers' experience) is our number one focus (and hence why we're taking a bit longer to get it right.) We still don't have a final design on how it works (hence why I hesitate to say "automatically") but I can assure you we are going to make it the best experience possible.

**Comment by marcobambini** [comment]
@nastiliano I assure you that 99% of the core offsync features will be fully transparent and automatic.

**Reply by nastiliano** [reply]
Thank you @marcobambini and @btholt .

But this is my dubt. How will it be work?

Will we need a local SQLlite Ddbb and when our apps start we call a PUB-SUB protocol?

Or we can use SQQLite Cloud directly???

I need a little bit more information because I've been waiting a lot and I already need to start to code.

Thank you very much

**Comment by marcobambini** [comment]
@nastiliano 
1. You open your local SQLite database
2. Load our extension
3. Call one function
4. Perform all read/write operations using the local SQLite database and the framework/language you like
5. Our extension/network layer will take care of syncing your local copy with all the peers connected with your cluster (with eventual consistency guarantee)

You can start developing your app with a local SQLite database, and then, with a couple of more lines of code, you'll have your offline sync solution when we release our extension.

**Reply by nastiliano** [reply]
ok, then...

I'm going to code my app as a regular "SQLite 3 API" and then, I'm going to join to your extra features (SQLite module)

Isn't it?

Thanks

**Comment by marcobambini** [comment]
@nastiliano yes, right

**Reply by nastiliano** [reply]
ok, thank you.

We'll see you on August/September 😃👍

Best regards

**Comment by nastiliano** [comment]
Hello everybody! @marcobambini @btholt 

Do we have any news?

We are near October and I'd like to know about progresses.

Thank you very much.

**Reply by nastiliano** [reply]
Hello?? @marcobambini @btholt

**Reply by jacobprall** [reply]
@nastiliano Hello! We are currently targeting a private beta of our local-first extension in early Q1. If you're interested in joining that private beta, please send an email to hello@sqlitecloud.io and you will be one of the first to know as soon as the extension is ready. 

Thank you for your patience and continued support!

**Comment by nastiliano** [comment]
I start to think that this project is just an illusion....

I'm still waiting for more than half a year...

Frustrating

**Reply by jacobprall** [reply]
Hello @nastiliano! We are still hard at work finalizing local-first. We are targeting a Q4 launch date. Thank you for your patience - we are laser-focused on providing the best possible local-first solution for our users!

**Reply by nastiliano** [reply]
Oh my God!
Finally!

Thank you very much

**Comment by helmysaeed** [comment]
Is there any update

**Reply by nastiliano** [reply]
I've seen there's a web interface update, but no "local-first" news yet :(

**Comment by nastiliano** [comment]
1 year since I post this topic and I think 1 year more waiting for this feature.

I already don't think in this project.

Best regards.
Good luck.

**Comment by marcobambini** [comment]
@nastiliano I completely understand your frustration, and I sincerely apologize for the delay.

At one point last year, we had to rewrite everything from the ground up. We took that as an opportunity to build a more robust and user-friendly solution—one that we believe gives us a real edge moving forward.

To give you a quick preview, I’ve attached a screenshot showing that everything is now ready for release. We’ve implemented cross-platform extensions, optional server-side row-level security, and updated the dashboard to support all these new features.

Thank you for your patience—we’re excited to share it in a few days.

![SQLiteSync2](https://github.com/user-attachments/assets/08c08dc4-2624-4d43-b762-7af9e0a1f20a)
![SQLiteSync](https://github.com/user-attachments/assets/404bb5f9-a9e2-4cc1-b9a4-36a0f6af3e9b)
![RLS](https://github.com/user-attachments/assets/7e6adfd3-bc28-4047-9cc5-6f0f5f0f49ad)

**Reply by nastiliano** [reply]
FINALLY!

I think I'm gonna cry :)

Best regards
